Pass mlflow.keras.load_model extra arguments to keras.models.load_model

yorickvP opened the pull request:
## What changes are proposed in this pull request?
 
Pass load_model arguments to keras to allow compile=False and custom_objects to be specified. Without this, it's impossible to load models with custom metrics. This fixes #1201.

Additionally, this adds a .split to the keras version check so it will work with "2.2.4-tf".
 
## How is this patch tested?
 
Running it in production.
 
## Release Notes
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)
It is now possible to pass extra arguments to mlflow.keras.load_model, that will be passed through to keras.load_model.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [x] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [x] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [x] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

yorickvP committed:
Pass mlflow.keras.load_model extra args to keras.models.load_model

Fixes #1201

yorickvP committed:
Handle keras versions like "2.2.4-tf"

yorickvP committed:
mlflow/keras.py: split documentation line

aarondav commented on mlflow/keras.py:
```diff
@@ -195,15 +195,16 @@ def _load_pyfunc(path):
 
 
-def load_model(model_uri):
+def load_model(model_uri, *args, **kwargs):
```
pytorch only passes along `**kwargs` here. This makes sense to me, since relying on positional params via a few layers of indirection seems a bit sketchy. Is there a strong reason to pass through `*args` as well?

yorickvP replied:
No.

aarondav commented on mlflow/keras.py:
```diff
@@ -146,21 +146,21 @@ def log_model(keras_model, artifact_path, conda_env=None, **kwargs):
 
 
-def _load_model(model_file):
+def _load_model(model_file, *args, **kwargs):
```
Could we add a unit test to `test_keras_model_export` that confirms that these extra kwargs are in fact passed through the keras properly on load?

yorickvP replied:
I don't know how to do this without mocking keras, since these arguments only make a difference when custom metrics are used.

aarondav commented:
Looks great, thanks! Just a couple of comments.

yorickvP committed:
Do not pass *args to keras load_model, only **kwargs

tomasatdatabricks commented on mlflow/keras.py:
```diff
@@ -177,7 +177,7 @@ def predict(self, dataframe):
 
 
-def _load_pyfunc(path):
+def _load_pyfunc(path, **kwargs):
```
We could also always just pass compile=False when loading as pyfunc?

tomasatdatabricks reviewed (commented):
Looks great thanks for the PR! 

I just left a comment regarding _load_pyfunc.

yorickvP committed:
Always pass compile=False from keras._load_pyfunc

Since load_pyfunc is used for prediction, and we don't need
to compile to do prediction, this will work in more cases.

tomasatdatabricks reviewed (approved):
Thanks for the updates. lgtm!
